test/unit/test/tmp_filename_test.cpp

```diff
@@ -67,13 +67,97 @@ TEST(tmp_filename_dtr, dtr)
     EXPECT_FALSE(fs::exists(path.parent_path()));
 }
 
-// throw if invalid TMPDIR
-TEST(tmp_filename_throw, throw)
+//!\brief A wrapper class to handle construction, destruction and permission handling of a temporary directory.
+class tmp_directory
 {
-    if (!system("man putenv > /dev/null 2>&1"))
+public:
+    tmp_directory() = delete; //< Deleted.
+    tmp_directory(tmp_directory const &) = delete; //< Deleted.
+    tmp_directory(tmp_directory &&) = default; //< Defaulted.
+    tmp_directory & operator=(tmp_directory const &) = delete; //< Deleted.
+    tmp_directory & operator=(tmp_directory &&) = default; //< Defaulted.
+
+    /*!\brief Construct a directory from a given path.
+     * \throws if std::filesystem::create_directory throws.
+     * \throws std::filesystem::filesystem_error if the directory already exists.
+     */
+    explicit tmp_directory(std::filesystem::path const & path) : directory_path(path)
+    {
+        if (!std::filesystem::create_directory(directory_path))
+        {
+            throw std::filesystem::filesystem_error{"The tmp_directory path already exists",
+                                                    directory_path,
+                                                    std::make_error_code(std::errc::file_exists)};
+        }
+    }
+
+    //!\brief Destructor tries to delete the directory.
+    ~tmp_directory()
     {
-        char str[] = "TMPDIR=/invalid";
-        putenv(str);
+        [[maybe_unused]] std::error_code ec;
+        // Add write permissions needed for deletion.
+        std::filesystem::permissions(directory_path,
+                                     std::filesystem::perms::owner_write,
+                                     std::filesystem::perm_options::add,
+                                     ec);
+        std::filesystem::remove(directory_path, ec);
```

Comment:
I would put this into a private function and share the code with the release function. Both call it with the error code but in addition the release function uses the error code to rethrow the exception if it was set.